### PR TITLE
[BEAM-13175] Add KinesisIO.write for AWS SDK v2.

### PR DESCRIPTION
**Pull request by mosche**
This implements the missing writer for Kinesis on AWS SDKV v2, one of the big milestones to get feature parity between the SDK v1 and v2 based IOs.

Currently the Kinesis Producer Library (KPL) is not available for SDK v2. Instead, this implements the writer using the async Kinesis client.

There's two different writer modes available:
- A default writer that asynchronously batches records to Kinesis with configurable concurrency.
- An aggregated record writer compatible with KPL / KCL record aggregation that is implemented on top of the default writer.

While record aggregation is compatible with KPL / KCL, aggregation here a lot less aggressive.
Only records with the same effective hash key are aggregated to keep complexity manageable. Otherwise, the aggregator would have to be aware of the most up-to-date explicit hash key ranges per shard and react promptly to resharding to not risk data being silently dropped on the consumer side.

https://docs.aws.amazon.com/streams/latest/dev/kinesis-kpl-concepts.html#kinesis-kpl-concepts-aggretation 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by mosche** [comment]
Run Java PreCommit

**Comment by mosche** [comment]
@mosche Don't forget to add Javadocs for the writer

**Comment by mosche** [comment]
R: @aromanenko-dev 
R: @echauchot 
R: @iemejia

**Comment by mosche** [comment]
Run Java PreCommit

**Comment by mosche** [comment]
Run Java PreCommit

**Comment by aromanenko-dev** [comment]
R: @lukecwik 
R: @chamikaramj 
Could you take a look if you have some time?

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java [hunk 1]
```diff
@@ -522,4 +651,393 @@ public Read withMaxCapacityPerShard(Integer maxCapacity) {
...
+      return builder().retryConfiguration(retryConfiguration).build();
+    }
+
+    public Write<T> validated() {
+      checkArgument(!isEmpty(streamName()), "streamName cannot be empty");
+      checkArgument(partitioner() != null, "partitioner cannot be null");
+      checkArgument(serializer() != null, "serializer cannot be null");
+      checkArgument(kinesisClientProvider() != null, "kinesisClientProvider cannot be null");
+      checkArgument(concurrentRequests() > 0, "concurrentRequests must be > 0");
+      checkArgument(
```
Comment (aromanenko-dev) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java:831]: Better to check it in `withBatchMaxRecords()`

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java [hunk 2]
```diff
@@ -522,4 +651,393 @@ public Read withMaxCapacityPerShard(Integer maxCapacity) {
...
+          batchMaxRecords() > 0 && batchMaxRecords() <= MAX_RECORDS_PER_REQUEST,
+          "batchMaxRecords must be in [1,%s]",
+          MAX_RECORDS_PER_REQUEST);
+      checkArgument(
```
Comment (aromanenko-dev) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java:835]: Better to check it in  `withBatchMaxBytes()`

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java [hunk 3]
```diff
@@ -522,4 +651,393 @@ public Read withMaxCapacityPerShard(Integer maxCapacity) {
...
+          batchMaxBytes() > 0 && batchMaxBytes() <= MAX_BYTES_PER_REQUEST,
+          "batchMaxBytes must be in [1,%s]",
+          MAX_BYTES_PER_REQUEST);
+      checkArgument(
```
Comment (aromanenko-dev) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java:839]: Better to check it in `aggregatedRecordsMaxBytes()`

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java [hunk 4]
```diff
@@ -522,4 +651,393 @@ public Read withMaxCapacityPerShard(Integer maxCapacity) {
...
+
+      private static ClientOverrideConfiguration clientOverride(RetryConfiguration retryConfig) {
+        return (retryConfig == null) ? null : retryConfig.toClientOverride();
+      }
+
+      protected final Write<T> spec;
+
+      private transient AsyncPutRecordsHandler handler;
+      private transient List<PutRecordsRequestEntry> requestEntries;
+      private transient int requestBytes;
```
Comment (aromanenko-dev) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java:875]: why this field is `transient`?

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java [hunk 5]
```diff
@@ -522,4 +651,393 @@ public Read withMaxCapacityPerShard(Integer maxCapacity) {
...
+        handler.reset();
+      }
+
+      @ProcessElement
+      public final void processElement(ProcessContext c) throws Throwable {
+        handler.checkForAsyncFailure();
+        T record = c.element();
+        byte[] data = spec.serializer().apply(record);
+        String pk = spec.partitioner().getPartitionKey(record);
+        String ehk = spec.partitioner().getExplicitHashKey(record);
```
Comment (aromanenko-dev) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java:907]: nit: `partitionKey` and `explicitHashKey` ?

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java [hunk 6]
```diff
@@ -522,4 +651,393 @@ public Read withMaxCapacityPerShard(Integer maxCapacity) {
...
+          MAX_RECORDS_PER_REQUEST);
+      checkArgument(
+          batchMaxBytes() > 0 && batchMaxBytes() <= MAX_BYTES_PER_REQUEST,
+          "batchMaxBytes must be in [1,%s]",
+          MAX_BYTES_PER_REQUEST);
+      checkArgument(
+          !aggregatedRecords() || aggregatedRecordsMaxBytes() <= MAX_BYTES_PER_RECORD,
+          "aggregatedRecordsMaxBytes must be <= %s",
+          MAX_BYTES_PER_RECORD);
+      checkArgument(
```
Comment (aromanenko-dev) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java:843]: nit: `checkState()`

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java [hunk 7]
```diff
@@ -522,4 +651,393 @@ public Read withMaxCapacityPerShard(Integer maxCapacity) {
...
+        }
+      }
+
+      @FinishBundle
+      public void finishBundle() throws Throwable {
+        flushEntries();
+        handler.waitForCompletion();
+      }
+
+      @Teardown
```
Comment (aromanenko-dev) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java:960]: `@Teardown` is a "best effort" to be called. Is it ok to rely on it in this case?

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java [hunk 8]
```diff
@@ -522,4 +651,393 @@ public Read withMaxCapacityPerShard(Integer maxCapacity) {
...
+      public void teardown() throws Exception {
+        clients.release(spec.kinesisClientProvider(), spec.retryConfiguration());
+      }
+
+      private void validatePartitionKey(String pk) {
```
Comment (aromanenko-dev) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java:965]: nit: `String partitionKey`

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java [hunk 9]
```diff
@@ -522,4 +651,393 @@ public Read withMaxCapacityPerShard(Integer maxCapacity) {
...
+        int size = pk != null ? pk.length() : 0;
+        checkArgument(
```
Comment (aromanenko-dev) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java:967]: nit: `checkState()`

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java [hunk 10]
```diff
@@ -522,4 +651,393 @@ public Read withMaxCapacityPerShard(Integer maxCapacity) {
...
+            PARTITION_KEY_MIN_LENGTH <= size && size <= PARTITION_KEY_MAX_LENGTH,
+            "Invalid partition key of length {}",
+            size);
+      }
+
+      private void validateExplicitHashKey(String hashKey) {
+        BigInteger key = new BigInteger(hashKey);
+        checkArgument(
```
Comment (aromanenko-dev) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java:975]: nit: checkState()

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/write/KinesisPartitioner.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+import java.io.Serializable;
+import java.math.BigInteger;
+import java.util.Random;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.apache.commons.lang3.StringUtils;
+
+/** Kinesis interface for custom partitioner. */
+public interface KinesisPartitioner<T> extends Serializable {
+  BigInteger MIN_HASH_KEY = new BigInteger("0");
```
Comment (aromanenko-dev) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/write/KinesisPartitioner.java:29]: Do we really need a BigInteger here?

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java [hunk 12]
```diff
@@ -522,4 +651,393 @@ public Read withMaxCapacityPerShard(Integer maxCapacity) {
...
+            key.compareTo(MIN_HASH_KEY) >= 0 && key.compareTo(MAX_HASH_KEY) <= 0,
+            "Explicit hash key must be 128-bit number.");
+      }
+    }
+
+    private static class AggregatedWriterFn<T> extends WriterFn<T> {
```
Comment (aromanenko-dev) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java:981]: Please, add a Javadoc with details of implementation for this class

**Review by aromanenko-dev — CHANGES_REQUESTED** [review]
Very demanded feature, thanks a lot @mosche for working on this! 

Due to the size of this PR, I did a first round of review over the main business logic (excluding tests).

Also, did you have a chance to run it against a real Kinesis instance and maybe compare the performance between this version and KPL's one (for AWS SDK v1)?

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java [hunk 13]
```diff
@@ -252,6 +371,16 @@ public static Read read() {
         .build();
   }
 
+  public static <T> Write<T> write() {
+    return new AutoValue_KinesisIO_Write.Builder<T>()
+        .batchMaxRecords(MAX_RECORDS_PER_REQUEST)
+        .batchMaxBytes((int) (MAX_BYTES_PER_REQUEST * 0.9)) // allow some error margin
+        .concurrentRequests(Write.DEFAULT_CONCURRENCY)
+        .aggregatedRecords(false)
```
[the remark this answers is not in the document]
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java:379]: Yes 👍

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java [hunk 14]
```diff
@@ -522,4 +651,393 @@ public Read withMaxCapacityPerShard(Integer maxCapacity) {
...
+      return withKinesisClientProvider(
+          new BasicKinesisProvider(credentialsProvider, region, endpoint));
+    }
+
+    /** Configure the AWS SDK retry behavior of the {@link KinesisAsyncClient}. */
+    public Write<T> withRetryConfiguration(RetryConfiguration retryConfiguration) {
+      return builder().retryConfiguration(retryConfiguration).build();
+    }
+
+    public Write<T> validated() {
```
[the remark this answers is not in the document]
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java:825]: Just needs to be visible for testing, will change 👍

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/write/KinesisPartitioner.java (hunk 11 above)
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/write/KinesisPartitioner.java:29]: Yes, it's 128 bits

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java (hunk 12 above)
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java:981]: 👍

**Comment by mosche** [comment]
Run Java PreCommit

**Comment by mosche** [comment]
> Also, did you have a chance to run it against a real Kinesis instance and maybe compare the performance between this version and KPL's one (for AWS SDK v1)?

I did quite some runs against real Kinesis using the integration test (using `--useLocalstack=false`). But I'll try and spend some time comparing performance. Though, the key issue will be the partitioner here. I'm not sure how to make it a meaningful & fair comparison. For instance, when using unique ids as partition key the new writer won't aggregate anything (except for some collisions on the effective hash key). On the other hand, using the provided random partitioner with the correct number of shards, it will aggregate as much as possible. The new writer certainly pushes a bit more responsibility to the user to think about partitioning upfront :/

**Comment by aromanenko-dev** [comment]
I think "a bit more responsibility to the user" is ok if it will help to achieve about the same performance as with KPL.

**Comment by mosche** [comment]
Run Java PreCommit

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java (hunk 3 above)
Comment (aromanenko-dev) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java]: ping

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java (hunk 4 above)
Comment (aromanenko-dev) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java:873]: ping

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java (hunk 5 above)
Comment (aromanenko-dev) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java]: ping

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java (hunk 6 above)
Comment (aromanenko-dev) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java:841]: ping

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java (hunk 7 above)
Comment (aromanenko-dev) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java:958]: ping

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java (hunk 3 above)
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java]: sorry, fixed already

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java (hunk 4 above)
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java:873]: Both, `requestEntries` and `requestBytes` are state that's only relevant / valid while processing a bundle.

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java (hunk 5 above)
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java]: fixed already

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java (hunk 7 above)
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java:958]: yes 👍

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java (hunk 6 above)
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java]: fixed 👍

**Comment by mosche** [comment]
Still planning to run a few more tests on EMR @aromanenko-dev

**Comment by mosche** [comment]
Blocked by https://github.com/apache/beam/pull/16760

**Comment by mosche** [comment]
Run Java PostCommit

**Comment by mosche** [comment]
@aromanenko-dev This should be in a good shape as an initial version. In my tests I was able to outperform the KPL based writer of SDK v1.

Using Spark I run into some interesting issues at high scale. These were caused by having classes of multiple Netty 4 versions on the classpath (see [here](https://github.com/aws/aws-sdk-java-v2/issues/1803)). It took me quite a while to figure this out as it only happened at very high throughput. 

I'm enforcing Beam's version of Netty over the more recent one used by AWS. In my tests this worked very well.

Another change is the output type of the writer. As recommended [here](https://docs.google.com/document/d/1V2FkGGunVgvLwi1dKHr-7mtDuwYjTuvuESV_oPzVnfQ/edit?resourcekey=0-KvfQq-5iCcMlu3f3MFJ-GQ#heading=h.xl97lw3dyot7) i changed it from `Void` to `Result` to allow for future backwards compatible changes. Currently the writer is fail fast (after retries of course), but adding a deadletter output would make sense...

The next step is to provide a more powerful internal partitioner that is aware of hashkey ranges assigned to KinesisShards.
With that record aggregation would be as powerful as the one provided by KPL.

cc @echauchot I'd be more than happy to get another review if you have the capacity for it ...

**Comment by mosche** [comment]
Run Java PreCommit

**Comment by aromanenko-dev** [comment]
@mosche Great progress, many thanks for details!

> I'm enforcing Beam's version of Netty over the more recent one used by AWS. In my tests this worked very well.

I think it should be fine.

> Another change is the output type of the writer. As recommended [here](https://docs.google.com/document/d/1V2FkGGunVgvLwi1dKHr-7mtDuwYjTuvuESV_oPzVnfQ/edit?resourcekey=0-KvfQq-5iCcMlu3f3MFJ-GQ#heading=h.xl97lw3dyot7) i changed it from Void to Result to allow for future backwards compatible changes. Currently the writer is fail fast (after retries of course), but adding a deadletter output would make sense...

Totally agree with that.

> The next step is to provide a more powerful internal partitioner that is aware of hashkey ranges assigned to KinesisShards.

Do you think it would make sense to do as a separate PR?

**Comment by mosche** [comment]
> Do you think it would make sense to do as a separate PR?

Absolutely, that's what I have in mind. I don't want to extend the scope of this one.

**Comment by mosche** [comment]
Run Java PreCommit

**Comment by mosche** [comment]
Run CommunityMetrics PreCommit

**Review by aromanenko-dev — APPROVED** [review]
LGTM
Great work on this and very important milestone! 
Thanks

**Pull request merged**